emathroughput.go

```diff
@@ -0,0 +1,359 @@
+type EMAThroughput struct {
+	// AdjustmentInterval defines how often we adjust the moving average from
+	// recent observations. Default 15s.
+	AdjustmentInterval time.Duration
+
+	// Weight is a value between (0, 1) indicating the weighting factor used to adjust
+	// the EMA. With larger values, newer data will influence the average more, and older
+	// values will be factored out more quickly.  In mathematical literature concerning EMA,
+	// this is referred to as the `alpha` constant.
+	// Default is 0.5
+	Weight float64
+
+	// InitialSampleRate is the sample rate to use during startup, before we
+	// have accumulated enough data to calculate a reasonable desired sample
+	// rate. This is mainly useful in situations where unsampled throughput is
+	// high enough to cause problems.
+	// Default 10.
+	InitialSampleRate int
+
+	// GoalThroughputPerSec is the target number of events to send per second.
+	// Sample rates are generated to squash the total throughput down to match the
+	// goal throughput. Actual throughput may exceed goal throughput. default 100
+	GoalThroughputPerSec int
+
+	// MaxKeys, if greater than 0, limits the number of distinct keys tracked in EMA.
+	// Once MaxKeys is reached, new keys will not be included in the sample rate map, but
+	// existing keys will continue to be be counted.
+	MaxKeys int
+
+	// AgeOutValue indicates the threshold for removing keys from the EMA. The EMA of any key will approach 0
+	// if it is not repeatedly observed, but will never truly reach it, so we have to decide what constitutes "zero".
+	// Keys with averages below this threshold will be removed from the EMA. Default is the same as Weight, as this prevents
+	// a key with the smallest integer value (1) from being aged out immediately. This value should generally be <= Weight,
+	// unless you have very specific reasons to set it higher.
+	AgeOutValue float64
+
+	// BurstMultiple, if set, is multiplied by the sum of the running average of counts to define
+	// the burst detection threshold. If total counts observed for a given interval exceed the threshold
+	// EMA is updated immediately, rather than waiting on the AdjustmentInterval.
+	// Defaults to 2; negative value disables. With a default of 2, if your traffic suddenly doubles,
+	// burst detection will kick in.
+	BurstMultiple float64
+
+	// BurstDetectionDelay indicates the number of intervals to run after Start is called before burst detection kicks in.
+	// Defaults to 3
+	BurstDetectionDelay uint
+
+	savedSampleRates map[string]int
+	currentCounts    map[string]float64
+	movingAverage    map[string]float64
+	burstThreshold   float64
+	currentBurstSum  float64
+	intervalCount    uint
+	burstSignal      chan struct{}
+
+	// haveData indicates that we have gotten a sample of traffic. Before we've
+	// gotten any samples of traffic, we should use the default goal sample rate
+	// for all events instead of sampling everything at 1
+	haveData bool
```

Comment:
Go has an atomic type that only allows setting a value once right?  Would that be useful here to ensure once it is set it can never get set back to false?

Reply:
This is a modified copy of the EMASampleRate sampler, which has this logic and text, as do many of the other samplers. I considered simply removing the feature -- I'm not sure it's worth the additional complexity -- but since it's in at least 4 samplers I decided to just leave it alone. I certainly don't think it's worth adding more complexity to avoid a purely theoretical possibility.